### PR TITLE
Add Combat Simulator

### PR DESCRIPTION
**Pull request by Morlic-fo**
This PR contains a Monte-Carlo style simulation for combat between two factions.
## SETUP

To use it, first check the file default/python/AI/SimulationShiplist.txt.
Make sure, the two lines
`FACTION_A` and `FACTION_B`
exist.
Any line under `FACTION_A` is interpreted as ship design name which is aligned with the first faction. If it doesn't exist or if it is spelled wrong, you will be informed when running the script.

Any line that comes under `FACTION_B` will be interpreted the same but added to the second faction.

So

```
FACTION_A
SD_MARK_1
SD_MARK_1
SD_MARK_1

FACTION_B
SD_AST_2
```

will be a battle between 3 SD_MARK_1 ship designs for faction A and a SD_AST_2 for faction B.

The design names used must be defined as usual by FOCS (see default/scripting/ship_designs)
## HOW TO RUN

1) Start Single Player Game with a single AI
2) write the command "simulate" into the game chat.
3) Read Simulation results in chat
Optional:
4) Edit SimulationShipList.txt to add/remove ships and save the file.
5) Run simulation again by writing the command "simulate" into the game chat. The new ships will be used for the simulation.
## EXAMPLE OUTPUT

```
AI_1: Adding Ships to Faction A now.
AI_1: Adding ship with name SD_MARK_1
AI_1: Adding ship with name SD_MARK_1
AI_1: Adding ship with name SD_MARK_1
AI_1: Adding Ships to Faction B now.
AI_1: Adding ship with name SD_AST_2
AI_1: Adding ship with name SD_AST_22
AI_1: ERROR: Could not find design.
AI_1: Number of ships A: 3 (cost: 140.4 PP)
AI_1: Number of ships B: 1 (cost: 118.6 PP)
AI_1: +++ SIMULATION RESULTS+++
AI_1: Simulated a total number of 150 iterations.
AI_1: Wins A: 0 (0.0%)
AI_1: Wins B: 88 (58.7%)
AI_1: Draws: 62 (41.3%)
AI_1: Losses Player A: (121.1 +- 23.1) PP
AI_1: Losses Player B: (0.0 +- 0.0) PP
```

The first lines are for debugging. Here, you can see that the design "SD_AST_22" was not found because there was a typo in it. The simulation was still performed with the remaining ships.

A summary of ship count and cost comparison is given (cost is influenced by game start fleet upkeep, not sure how to avoid this) followed by the simulation results.
A Draw either means that both fleets are destroyed or that both fleets survived.
The losses are calculated by destroyed ships. The first number are the average losses and the second number is the simulated standard deviation.
## Further comments
- Designs and parts are queried directly from the server
- Currently, no tech/species updates are considered yet.
- If you change content (FOCS) scripts, you need to start a new game for the changes to take effect.
- Framework for planets exists but has no textfile interface yet.

### CODE REVIEW

**Comment by MatGB** [comment]
Ooh, interesting.

Almost certainly going to be useful for balancing stuff, trying to get specific types of design into fights is always a pain. Thank you.

File: default/python/AI/combat_simulator.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+                results.append(WIN_B)
+            else:
+                results.append(DRAW)
+
+        mean_losses_A = mean(lost_pp_A)
+        std_losses_A = std(lost_pp_A)
+        mean_losses_B = mean(lost_pp_B)
+        std_losses_B = std(lost_pp_B)
+
+        draws = len([r for r in results if r == DRAW])
```
Comment (Cjkjvfnby) [inline default/python/AI/combat_simulator.py:139]: `len(r for r in results if r == DRAW)`  # generator expression
`sum(r == DRAW for r in results)`

File: default/python/AI/combat_simulator.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        wins_a = len([r for r in results if r == WIN_A])
+        wins_b = len([r for r in results if r == WIN_B])
+
+        chat_human("+++ SIMULATION RESULTS+++")
+        chat_human("Simulated a total number of %d iterations." % iterations)
+        chat_human("Wins A: %d (%.1f%%)" % (wins_a, 100*float(wins_a)/iterations))
```
Comment (Cjkjvfnby) [inline default/python/AI/combat_simulator.py:145]: `100.0 * wins_a / iterations`

**Comment by LGM-Doyle** [comment]
Awesome.  

It will make the AI more clever by using actual game mechanics to predict combat outcome, but you already knew that.

I like this idea.

**Comment by Vezzra** [comment]
@adrianbroher, has closing this issue been intentional, and if yes, why did you close it? You neither provided any comment, nor set an appropriate "status:*" label...

**Comment by adrianbroher** [comment]
@Vezzra, I don't remember interacting with this Issue let alone closing it.

**Comment by Vezzra** [comment]
Ok, just wanted to check back with you before reopening.

However, just now, when I wanted to reopen, I noticed that apparently the reason for this having been closed is because the Fighters branch got deleted (after it had been merged). As you've been the one who deleted the Fighters branch, you show up as the one closing this issue.

So that's cleared up now. @Morlic-fo: can you fix this PR (probably by rebasing)? Or, if you don't intend to maintain it anymore, set an appropriate "status:*" label.

**Comment by MatGB** [comment]
Observation: this was a PR to the Fighters branch, which no longer exists. I suspect I forgot to delete that branch on merge and @adrianbroher did so, hence it'll show him as closing this.

Given this was always a WIP and not intended for merge it's fine, but we definitely want to keep it as I need to do a proper costs balancing pass at some point nearer the end of this cycle (internal slots are currently undervalued as they did nothing last time I did a pass so didn't cost them)

**Comment by Morlic-fo** [comment]
Opened new PR #1175

**Comment by Vezzra** [comment]
Thanks!

**Pull request closed without merging**